### ISSUE TITLE
Use BSBGateway with BSB-LAN module connected to Fujitsu Waterstage

### ISSUE THREAD

**Issue by Lamarqe** [loehnertj/bsbgateway#18]
I try to use BSBGateway in combination with a BSB-LAN module as described here: 
https://1coderookie.github.io/BSB-LPB-LAN/kap01.html

Currently without success. (No data is received even though the adapter signals a proper connection to the BSB interface)

I have two questions regarding the configuration in config.py:

- According to the manual invert_bytes and port_parity need to be set to True/odd in case of a "simple BSB-to-UART level converter". Otherwise to False/even. 
Is the assumption correct that the linked BSB-LAN module is a more sophisticated adapter that does not require inverting of bits? Meaning that these parameters need to be set to False/even in this setup?

- What is the impact of the statement: "Currently there is only a (incomplete) driver for Broetje ISR Plus." I own a Fujitsu Waterstage, not a Broetje device. Should it still be possible to receive data using BSBGateway in the default config? Once the general connectivitiy works and I receive just any data, I am happy to contribute with input for this heatpump.

**Comment by larspelz** [comment]
Hi Lamarge,
I have a bsblan interface and bsbgateway running happily on a Raspberry Pi Zero. I do own a Broetje gas boiler, but you should at least receive some data from the bus as your [Waterstage should be compatible with the BSB-LAN adapter. ](https://mattstech.info/posts/fujitsu-waterstage-domotica/)
In my config.py I changed
`'adapter_device': '/dev/ttyUSB0'`
to
`'adapter_device': '/dev/ttyAMA0'`
I also commented out the lines 
`data = str(data)`
in bsbgateway/serial_source.py but that seemed to be a problem with my python version.
Hope that helps.

**Comment by Lamarqe** [comment]
That means you left invert_bytes and port_parity at True/odd and this worked with the BSB-LAN module?
Thank you for that information! This reduces a lot of uncertainity and will help me to further narrow down "my" problem.

Looks like the string conversion issue has been fixed with the [most recent commit](https://github.com/loehnertj/bsbgateway/commit/c348bcd478237c3a6f51d8bd0b9ef1a96deb3cd9). So I have this fix already.

**Comment by loehnertj** [comment]
Hi Lamarqe, thanks for your interest! As lars wrote, **invert_bytes=True** and **port_parity=odd** should be the right setting for the BSB LAN and actually also for 99% of all other circuits. (this has to do with the different definitions of "0" and "1" voltage levels for BSB vs. RS232 standard).

String conversion issue is indeed fixed by the recent commit.

As for your second question. You should at least be able to read received data and have it split it up into sender, receiver, type and payload fields. The program might not recognize the field and payload datatype though.

If you turn "dump" on in the console interface and play a bit with the heater's builtin control panel, you will see messages like this for known fields:
``<BsbTelegram 0 -> 24: unknown (8) 8337 Startzähler Brenner = [7] [raw:07 ] @1452418743.752471>``

and like this for unknown fields:
``<BsbTelegram 0 -> 127: inf 0x2F000211 = [0, 0, 255, 255, 255, 255, 255, 255, 0, 0]>``

(notice the name and 4-digit ID number in the first line, opposed to the 8-digit hex id in second line).

The current builtin field list contains some fields from the early days of reverse engineering this stuff, as well as some things sniffed off my own heater. I am working on getting the field list from the BSB-Lan project imported, which is vastly more complete. But this is not ready yet :-(.

**Comment by Lamarqe** [comment]
Connectivity issues appeared to be a result of inverted RX/TX pins. 😞 

Having that fixed, I run into exceptions, even before triggering the dump command:

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.9/threading.py", line 892, in run
    self._target(*self._args, **self._kwargs)
  File "/home/scritch/bsbgateway/bsbgateway/event_sources.py", line 86, in _run_wrapper
    o.run(putevent_func)
  File "/home/scritch/bsbgateway/bsbgateway/bsb/bsb_comm.py", line 80, in run
    o.serial.run(convert_data)
  File "/home/scritch/bsbgateway/bsbgateway/serial_source.py", line 103, in run
    o.serial_port = serial.Serial(**o._serial_arg)
  File "/usr/lib/python3/dist-packages/serial/serialutil.py", line 244, in __init__
    self.open()
  File "/usr/lib/python3/dist-packages/serial/serialposix.py", line 329, in open
    self._reconfigure_port(force_update=True)
  File "/usr/lib/python3/dist-packages/serial/serialposix.py", line 499, in _reconfigure_port
    termios.tcsetattr(
termios.error: (22, 'Invalid argument')
INFO bsbgateway.web_interface:84 @1049 -- Web interface listening on http://0.0.0.0:8081/
ERROR bsbgateway.bsb_gateway:87 @4796 -- Something crashed while processing event timer with data None
Traceback (most recent call last):
  File "/home/scritch/bsbgateway/bsbgateway/bsb_gateway.py", line 85, in _dispatch_event
    getattr(o, 'on_%s_event'%evtype)(evdata)
  File "/home/scritch/bsbgateway/bsbgateway/bsb_gateway.py", line 93, in on_timer_event
    logger.tick()
  File "/home/scritch/bsbgateway/bsbgateway/single_field_logger.py", line 89, in tick
    o.send_get_telegram(o.field.disp_id)
  File "/home/scritch/bsbgateway/bsbgateway/bsb_gateway.py", line 57, in <lambda>
    logger.send_get_telegram = lambda disp_id: o._bsbcomm.send_get(disp_id)
  File "/home/scritch/bsbgateway/bsbgateway/bsb/bsb_comm.py", line 128, in send_get
    o.serial.write(t.serialize())
  File "/home/scritch/bsbgateway/bsbgateway/serial_source.py", line 145, in write
    o.serial_port.write(data)
AttributeError: 'NoneType' object has no attribute 'write'
```

These are two exceptions:
- The first one appears immediately after starting bsbgateway.
- The second one appears cyclically, roughly every second

**Comment by loehnertj** [comment]
Seems to be another Python2->3 related bug. I just pushed a change that will hopefully fix it (5e7ced1). Please try and report.

**Comment by Lamarqe** [comment]
Yes, the fix works well. Thank you.
Command line and web based access works now!

I started playing around with the REST calls with the intention to integrate the values into home assistant.
Here, I ran into an issue during the json serialization:

Eg, using the command line wget http://127.0.0.1:8081/field-8700.value

Lead to the following exception:
```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/web/application.py", line 283, in process
    return self.handle()
  File "/usr/lib/python3/dist-packages/web/application.py", line 274, in handle
    return self._delegate(fn, self.fvars, args)
  File "/usr/lib/python3/dist-packages/web/application.py", line 502, in _delegate
    return handle_class(f)
  File "/usr/lib/python3/dist-packages/web/application.py", line 495, in handle_class
    return tocall(*args)
  File "/home/scritch/bsbgateway/bsbgateway/web_interface/field.py", line 51, in GET
    return bridge_call(o, view, q, o.parameters)
  File "/home/scritch/bsbgateway/bsbgateway/web_interface/webutils.py", line 107, in bridge_call
    result = autojson(result)
  File "/home/scritch/bsbgateway/bsbgateway/web_interface/webutils.py", line 164, in autojson
    result = json.dumps(obj)
  File "/usr/lib/python3.9/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.9/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.9/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bytes is not JSON serializable

INFO bsbgateway.web_interface:144 @451138 -- 127.0.0.1:57632: GET /field-8700.value - 500 Internal Server Error

```

What I did was the following change:

```
diff --git a/bsbgateway/web_interface/webutils.py b/bsbgateway/web_interface/webutils.py
index 0e17fd3..a384044 100644
--- a/bsbgateway/web_interface/webutils.py
+++ b/bsbgateway/web_interface/webutils.py
@@ -161,7 +161,7 @@ def autojson(obj):
     if hasattr(obj, 'to_json'):
         result = obj.to_json()
     else:
-        result = json.dumps(obj)
+        return obj
     web.header('Content-Type', 'application/json; charset=utf-8')
     return result

```
This works, but I don't know if this is a proper fix. (I am not experienced in python, would have to dig more into it).

**Comment by loehnertj** [comment]
I am amazed that your change works. :-) I now pushed a proper fix for it.

On the topic of the fields. The BSB-Lan "master" data table is a giant C header file, which is maybe somewhat hard to read. (e.g. it uses variables like STR_0815 as field name, in order to redirect to language-specific string tables). This is available in their Github repo - link below.

From the maintainer, I got a JSON-conversion of that data in early 2022. The JSON file is apparently is not online anywhere. I can send it to you in private if it is helpful.

*However*, (and this is new to me too as of today):  The BSB-LAN guys decided to strip down their table in September 2022. Reasons are explained very thoroughly here: https://github.com/fredlcore/BSB-LAN/discussions/482
Summarized in one sentence - they decided that further on there will be only _device-specific_ parameter tables handed out on request, because there are numerous and in some cases dangerous differences between parameter sets of individual heaters. So be warned. I yet have to make up my mind on how I will proceed with this new situation.

Anyway, the full field table is of course still available in the Github history: https://raw.githubusercontent.com/fredlcore/BSB-LAN/01e6b4fac52f3f4f880c610bf3ad9764dd766d81/BSB_LAN/BSB_LAN_defs.h
Actual field list starts somewhere around the middle; search for "global command table".

As I said, I can also send you the JSON version of the same data.

**Comment by Lamarqe** [comment]
Thank you for the insights. I would be highly interested in the json version. That would be great to have.
I am not concerned about wrong parameter sets, as I intend to set up just a few dozen values in monitoring way, without sending commands. If any, I will cross check with the official value sets (as I said, I have the full documentation of my device including all paramter IDs and value names at hand).
Could you send the file? (I guess via public file share service?)

Btw: The json fix works. Great!
I still have some issues related with requests which are not getting a response. Eg when I use the "load all" button on the default config of group-0. Loading stops after having received a value for key 658. Afterwards, key 5970 is not getting processed. 

When I then try to leave bsbgateway using "quit", the process does not exit. Only after using multiple Ctrl+C, the final event source stops ("event source hub "). I can reproduce this behavior deterministically. 
```
INFO bsbgateway.web_interface:144 @37087 -- 192.168.2.51:22327: GET /field-1014.widget - 200 OK
INFO bsbgateway.web_interface:144 @37586 -- 192.168.2.51:22327: GET /field-1012.widget - 200 OK
INFO bsbgateway.web_interface:144 @38103 -- 192.168.2.51:22327: GET /field-1010.widget - 200 OK
INFO bsbgateway.web_interface:144 @38587 -- 192.168.2.51:22327: GET /field-658.widget - 200 OK
quit
DEBUG bsbgateway.event_sources:105 @195922 -- event source  will stop somewhen...
DEBUG bsbgateway.event_sources:105 @195923 -- event source  will stop somewhen...
DEBUG bsbgateway.event_sources:93 @195923 -- event source web exited.
DEBUG bsbgateway.event_sources:93 @196209 -- event source timer exited.
^C^C^C^C^C^CDEBUG bsbgateway.event_sources:93 @234390 -- event source hub exited.
```

When I then try to repdroduce the problem again, (after restart) the problem re-appears earlier again, until I finally cannot get any values from this client system anymore. 😮 . Feels to me like TCP/IP connections which are not getting closed between bsbgateway and the client.
I believe this is related with using the WebUI. Just using wget .value requests, the problem has not yet shown up for me.
As I will just access the .value URLs in the productive setup, I currently hope&expect that this problem will not affect my final usage.

**Comment by loehnertj** [comment]
Apparantly you can't send direct messages over github. Send me a mail (loehnert.kde@gmx.de) and I'll reply with the files attached. I don't really want to post them in public yet.

Your analysis of the missing timeout is probably correct. I'll take care of it but maybe not this weekend.

**Comment by Lamarqe** [comment]
Thank you! We can close this issue at this point, as the initial setup questions have been solved and the setup is working fine.

There are a two additional problems, which are not specific to the usage of a fujitsu waterstage.
